crates/sui-core/src/authority/shared_object_version_manager.rs

```diff
@@ -151,11 +152,14 @@ impl SharedObjVerManager {
         let receiving_object_keys = transaction_receiving_object_keys(cert);
         input_object_keys.extend(receiving_object_keys);
 
-        if let Some(congested_objects) = cancellation_info {
+        if txn_cancelled {
             // For cancelled transaction due to congestion, assign special versions to all shared objects.
             // Note that new lamport version does not depend on any shared objects.
             for SharedInputObject { id, .. } in shared_input_objects.iter() {
-                let assigned_version = if congested_objects.contains(id) {
+                let assigned_version = if congested_objects_info
+                    .as_ref()
+                    .is_some_and(|info| info.contains(id))
+                {
                     SequenceNumber::CONGESTED
                 } else {
                     SequenceNumber::CANCELLED_READ
```

Comment:
If shared objects are assigned to `SequenceNumber::CANCELLED_READ`, this assertion will be triggered: https://github.com/MystenLabs/sui/blob/98f6e1acddbb87ca0a7e337099e7fd1d695f0d2d/crates/sui-types/src/transaction.rs#L3032

Currently, the cancellation logic assumes that when execution sees `CANCELLED_READ`, it assumes that the transaction is cancelled due to shared object congestion control.